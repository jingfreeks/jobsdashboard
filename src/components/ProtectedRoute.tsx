import { ReactNode } from 'react';
import { Navigate } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { selectIsAuthenticated } from '@/features/auth';
import type { RootState } from '@/config/store';

interface ProtectedRouteProps {
  children: ReactNode;
}

const ProtectedRoute = ({ children }: ProtectedRouteProps) => {
  const isAuthenticated = useSelector((state: RootState) => selectIsAuthenticated(state));

  if (!isAuthenticated) {
    // Redirect to login if user is not authenticated
    return <Navigate to="/login" replace />;
  }

  // Render the protected component if user is authenticated
  return <>{children}</>;
};

export default ProtectedRoute; 