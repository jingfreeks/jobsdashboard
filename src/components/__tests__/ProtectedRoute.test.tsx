import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import authReducer from '@/features/auth';
import ProtectedRoute from '../ProtectedRoute';
import type { RootState } from '@/config/store';

// Mock component to render inside ProtectedRoute
const MockComponent = () => <div>Protected Content</div>;

const createTestStore = (initialAuthState: RootState['auth']) => {
  return configureStore({
    reducer: {
      auth: authReducer,
    },
    preloadedState: {
      auth: initialAuthState,
    },
  });
};

const renderWithProviders = (component: React.ReactElement, store: ReturnType<typeof configureStore>) => {
  return render(
    <Provider store={store}>
      <BrowserRouter>
        {component}
      </BrowserRouter>
    </Provider>
  );
};

describe('ProtectedRoute', () => {
  it('should render children when user is authenticated', () => {
    const store = createTestStore({
      user: 'testuser',
      token: 'valid-token',
      userId: '123',
      roles: ['user'],
    });

    const { container } = renderWithProviders(
      <ProtectedRoute>
        <MockComponent />
      </ProtectedRoute>,
      store
    );

    expect(screen.getByText('Protected Content')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should redirect to login when user is not authenticated', () => {
    const store = createTestStore({
      user: null,
      token: null,
      userId: null,
      roles: [],
    });

    const { container } = renderWithProviders(
      <ProtectedRoute>
        <MockComponent />
      </ProtectedRoute>,
      store
    );

    // The component should not render the protected content
    expect(screen.queryByText('Protected Content')).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('matches protected route snapshot when authenticated', () => {
    const store = createTestStore({
      user: 'testuser',
      token: 'valid-token',
      userId: '123',
      roles: ['user'],
    });

    const { container } = renderWithProviders(
      <ProtectedRoute>
        <MockComponent />
      </ProtectedRoute>,
      store
    );

    expect(container).toMatchSnapshot();
  });

  it('matches protected route snapshot when not authenticated', () => {
    const store = createTestStore({
      user: null,
      token: null,
      userId: null,
      roles: [],
    });

    const { container } = renderWithProviders(
      <ProtectedRoute>
        <MockComponent />
      </ProtectedRoute>,
      store
    );

    expect(container).toMatchSnapshot();
  });
}); 