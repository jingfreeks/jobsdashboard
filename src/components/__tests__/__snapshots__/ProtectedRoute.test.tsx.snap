// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProtectedRoute > matches protected route snapshot when authenticated 1`] = `
<div>
  <div>
    Protected Content
  </div>
</div>
`;

exports[`ProtectedRoute > matches protected route snapshot when not authenticated 1`] = `<div />`;

exports[`ProtectedRoute > should redirect to login when user is not authenticated 1`] = `<div />`;

exports[`ProtectedRoute > should render children when user is authenticated 1`] = `
<div>
  <div>
    Protected Content
  </div>
</div>
`;
